src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1368,52 +1432,363 @@ private sealed class RemainingTestsForCase
+                    out Tests whenTrue,
+                    out Tests whenFalse,
+                    ref bool foundExplicitNullTest)
+                {
+                    Negated.Filter(builder, test, whenTrueValues, whenFalseValues, out var whenTestTrue, out var whenTestFalse, ref foundExplicitNullTest);
+                    whenTrue = Not.Create(whenTestTrue);
+                    whenFalse = Not.Create(whenTestFalse);
                 }
+                public override bool Equals(object? obj) => obj is Not n && Negated.Equals(n.Negated);
+                public override int GetHashCode() => Hash.Combine(Negated.GetHashCode(), typeof(Not).GetHashCode());
+            }
 
-                return result;
+            public abstract class SequenceTests : Tests
+            {
+                public readonly ImmutableArray<Tests> RemainingTests;
+                protected SequenceTests(ImmutableArray<Tests> remainingTests)
+                {
+                    Debug.Assert(remainingTests.Length > 1);
+                    this.RemainingTests = remainingTests;
+                }
+                public abstract Tests Update(ArrayBuilder<Tests> remainingTests);
+                public override void Filter(
+                    DecisionDagBuilder builder,
+                    BoundDagTest test,
+                    IValueSet? whenTrueValues,
+                    IValueSet? whenFalseValues,
+                    out Tests whenTrue,
+                    out Tests whenFalse,
+                    ref bool foundExplicitNullTest)
+                {
+                    var trueBuilder = ArrayBuilder<Tests>.GetInstance(RemainingTests.Length);
+                    var falseBuilder = ArrayBuilder<Tests>.GetInstance(RemainingTests.Length);
+                    foreach (var other in RemainingTests)
+                    {
+                        other.Filter(builder, test, whenTrueValues, whenFalseValues, out Tests oneTrue, out Tests oneFalse, ref foundExplicitNullTest);
+                        trueBuilder.Add(oneTrue);
+                        falseBuilder.Add(oneFalse);
+                    }
+
+                    whenTrue = Update(trueBuilder);
+                    whenFalse = Update(falseBuilder);
+                }
+                public override Tests RemoveEvaluation(BoundDagEvaluation e)
+                {
+                    var builder = ArrayBuilder<Tests>.GetInstance(RemainingTests.Length);
+                    foreach (var test in RemainingTests)
+                        builder.Add(test.RemoveEvaluation(e));
+
+                    return Update(builder);
+                }
+                public override bool Equals(object? obj) =>
+                    obj is SequenceTests other && this.GetType() == other.GetType() && RemainingTests.SequenceEqual(other.RemainingTests);
+                public override int GetHashCode()
+                {
+                    int length = this.RemainingTests.Length;
```

Comment:
>length [](start = 24, length = 6)

`length` is not necessary since the hash codes of all the elements are combined.

Reply:
I don't know enough about the implementation of `Hash.Combine` to know if this helps (with hash value distribution) or not. In any case, I don't believe it is harmful, so I'm going to leave it.

---
In reply to: [382888446](https://github.com/dotnet/roslyn/pull/41503#discussion_r382888446) [](ancestors = 382888446)